controllers/grafanaalertrulegroup_controller.go

```diff
@@ -0,0 +1,371 @@
+	if err != nil && !errors.As(err, &ruleNotFound) {
+		return fmt.Errorf("fetching existing alert rule group: %w", err)
+	}
+
+	currentRules := make(map[string]bool)
+	if applied != nil {
+		for _, rule := range applied.Payload.Rules {
+			currentRules[rule.UID] = false
+		}
+	}
+
+	for _, rule := range group.Spec.Rules {
+		rule := rule
+		apiRule := &models.ProvisionedAlertRule{
+			Annotations:  rule.Annotations,
+			Condition:    &rule.Condition,
+			Data:         make([]*models.AlertQuery, len(rule.Data)),
+			ExecErrState: &rule.ExecErrState,
+			FolderUID:    &folderUID,
+			For:          (*strfmt.Duration)(&rule.For.Duration),
+			IsPaused:     rule.IsPaused,
+			Labels:       rule.Labels,
+			NoDataState:  rule.NoDataState,
+			RuleGroup:    &group.Name,
+			Title:        &rule.Title,
+			UID:          rule.UID,
+		}
+		for idx, q := range rule.Data {
+			apiRule.Data[idx] = &models.AlertQuery{
+				DatasourceUID:     q.DatasourceUID,
+				Model:             q.Model,
+				QueryType:         q.QueryType,
+				RefID:             q.RefID,
+				RelativeTimeRange: q.RelativeTimeRange,
+			}
+		}
+
+		if _, ok := currentRules[rule.UID]; ok {
+			params := provisioning.NewPutAlertRuleParams().
+				WithBody(apiRule).
+				WithXDisableProvenance(&strue).
+				WithUID(rule.UID)
+			_, err := cl.Provisioning.PutAlertRule(params) //nolint:errcheck
+			if err != nil {
+				return fmt.Errorf("updating rule: %w", err)
+			}
+		} else {
+			params := provisioning.NewPostAlertRuleParams().
+				WithBody(apiRule).
+				WithXDisableProvenance(&strue)
+			_, err = cl.Provisioning.PostAlertRule(params) //nolint:errcheck
+			if err != nil {
+				return fmt.Errorf("creating rule: %w", err)
+			}
+		}
+
+		currentRules[rule.UID] = true
+	}
+
+	for uid, present := range currentRules {
+		if !present {
+			params := provisioning.NewDeleteAlertRuleParams().
+				WithUID(uid).
+				WithXDisableProvenance(&strue)
+			_, err := cl.Provisioning.DeleteAlertRule(params) //nolint:errcheck
+			if err != nil {
+				return fmt.Errorf("deleting old alert rule %s: %w", uid, err)
+			}
+		}
+	}
+
+	mGroup := &models.AlertRuleGroup{
+		FolderUID: folderUID,
+		Interval:  int64(group.Spec.Interval.Seconds()),
+		Rules:     []*models.ProvisionedAlertRule{},
+		Title:     "",
+	}
+	params := provisioning.NewPutAlertRuleGroupParams().
+		WithBody(mGroup).
+		WithGroup(group.Name).
+		WithFolderUID(folderUID).
+		WithXDisableProvenance(&strue)
+	_, err = cl.Provisioning.PutAlertRuleGroup(params) //nolint:errcheck
+	if err != nil {
+		return fmt.Errorf("updating group: %s", err.Error())
+	}
+	return nil
+}
+
+func (r *GrafanaAlertRuleGroupReconciler) finalize(ctx context.Context, group *grafanav1beta1.GrafanaAlertRuleGroup) error {
+	folderUID := r.GetFolderUID(ctx, group)
+	if folderUID == "" {
+		r.Log.Info("ignoring finalization logic as folder no longer exists")
+		return nil
+	}
+	instances, err := r.GetMatchingInstances(ctx, group, r.Client)
```

Comment:
If we don't got any instances I think we should be safe to remove the finalizer.

Something like this

```go
if instances == nil {
    r.Log.Info("ignoring finalization logic as there is no grafana instance match)
    return nil
}
```